src/Psl/Shell/execute.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Shell;
+
+use Psl\Dict;
+use Psl\Env;
+use Psl\Str;
+use Psl\Vec;
+
+use function fclose;
+use function is_dir;
+use function is_resource;
+use function proc_close;
+use function proc_open;
+use function stream_get_contents;
+
+/**
+ * Execute an external program.
+ *
+ * @param string                $command            The command to execute.
+ * @param list<string>          $arguments          The command arguments listed as separate entries.
+ * @param string                $working_directory  The initial working directory for the command.
+ *                                                  This must be an absolute directory path, or null
+ *                                                  if you want to use the default value (the working dir of the
+ *                                                  current directory )
+ * @param array<string, string> $environment        An array with the environment variables for the command that
+ *                                                  will be run.
+ * @param bool                  $escape_arguments   If set to true ( default ), all $arguments will be escaped using
+ *                                                  `Shell\escape_argument`.
+ *
+ * @psalm-taint-sink shell $command
+ *
+ * @throws Exception\FailedExecutionException   In case the command resulted in an exit code other than 0.
+ * @throws Exception\PossibleAttackException    In case the command being run is suspicious ( e.g: contains NULL byte ).
+ * @throws Exception\RuntimeException           In case $working_directory doesn't exist, or unable to create a new
+ *                                              process.
+ */
+function execute(
+    string $command,
+    array $arguments = [],
+    ?string $working_directory = null,
+    array $environment = [],
+    bool $escape_arguments = true
+): string {
```

Comment:
`shell_exec` and `exec` are not powerful enough, `proc_*` is a bit complex, i think this command is a nice middle ground where you can use a costume working directory, specify environment variables, and have more control ( e.g: `exec` will redirect error output of the command being ran to PHPs own process STDERR, which is not what you want most likely, but using `Shell\execute`, you get stderr output in the exception in case of an error ).

However, `Shell\execute` is still not as powerful as `proc_*` ( e.g: cannot write to process input ), so i guess if people want more control, they can either use proc* or `symfony/process` package.



WDYT @veewee?

Reply:
Nice work! For most of the cases, this is sufficient. `proc_` surely is the way to go!
Maybe we can also provide a stdin ?ReadHandle param for writing to the stdin at some point that streams to pipe[0].

Can't wait to drop my `symfony/process` dependencies!

Reply:
as i mentioned below, we don't want anything more complex until #127 is merged.